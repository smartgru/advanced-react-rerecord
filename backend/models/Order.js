import { Text, Integer, Relationship, DateTime } from '@keystonejs/fields';

import { byTracking, atTracking } from '@keystonejs/list-plugins';

export default {
  labelResolver: item => `Sale ${item.length}`,
  fields: {
    total: { type: Integer },
    items: {
      type: Relationship,
      ref: 'OrderItem',
      many: true,
    },
    user: {
      type: Relationship,
      ref: 'User',
    },
    charge: { type: Text },
    // TODO: Wait for virtual fields to be published
    // display_charge: {
    //   type: Virtual,
    //   resolver: item => `Charge # ${item.charge}`,
    // },
    createdAt: {
      type: DateTime,
      format: 'MM/DD/YYYY h:mm A',
      yearRangeFrom: 1901,
      yearRangeTo: 2018,
      yearPickerType: 'auto',
    },
  },
  plugins: [atTracking(), byTracking()],
};
