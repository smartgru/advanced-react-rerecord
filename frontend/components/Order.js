import React from 'react';
import PropTypes from 'prop-types';
import { Query } from '@apollo/client';
import { format } from 'date-fns';
import Head from 'next/head';
import gql from 'graphql-tag';
import { useQuery } from '@apollo/client';
import formatMoney from '../lib/formatMoney';
import Error from './ErrorMessage';
import OrderStyles from './styles/OrderStyles';

const SINGLE_ORDER_QUERY = gql`
  query SINGLE_ORDER_QUERY($id: ID!) {
    Order(where: { id: $id }) {
      id
      charge
      total
      createdAt
      user {
        id
      }
      items {
        id
        name
        description
        price
        image
        quantity
      }
    }
  }
`;

function Order({ id }) {
  const { data, error, loading } = useQuery(SINGLE_ORDER_QUERY, {
    variables: { id },
  });
  if (error) return <Error error={error} />;
  if (loading) return <p>Loading...</p>;
  const { Order: order } = data;
  return (
    <OrderStyles data-test="order">
      <Head>
        <title>Sick Fits - Order {order.id}</title>
      </Head>
      <p>
        <span>Order ID:</span>
        <span>{id}</span>
      </p>
      <p>
        <span>Charge</span>
        <span>{order.charge}</span>
      </p>
      <p>
        <span>Date</span>
        <span>
          {format(new Date(order.createdAt), 'MMMM d, yyyy h:mm a', {
            awareOfUnicodeTokens: true,
          })}
        </span>
      </p>
      <p>
        <span>Order Total</span>
        <span>{formatMoney(order.total)}</span>
      </p>
      <p>
        <span>Item Count</span>
        <span>{order.items.length}</span>
      </p>
      <div className="items">
        {order.items.map(item => (
          <div className="order-item" key={item.id}>
            <img src={item.image} alt={item.title} />
            <div className="item-details">
              <h2>{item.name}</h2>
              <p>Qty: {item.quantity}</p>
              <p>Each: {formatMoney(item.price)}</p>
              <p>SubTotal: {formatMoney(item.price * item.quantity)}</p>
              <p>{item.description}</p>
            </div>
          </div>
        ))}
      </div>
    </OrderStyles>
  );
}

Order.propTypes = {
  id: PropTypes.string.isRequired,
};

export default Order;
export { SINGLE_ORDER_QUERY };
