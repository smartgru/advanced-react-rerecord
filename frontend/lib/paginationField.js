export default function paginationField() {
  return {
    keyArgs: false, // take full control of this field
    // We write custom functions to merge and read based on the 'first' and 'skip' args
    // first = 4 items per page
    // skip = how many items to offset
    merge(existing, incoming, { args }) {
      const { skip, first } = args;
      console.log('merging...', { first, skip, existing, incoming });

      const merged = existing ? existing.slice(0) : [];
      // we do it this way because someone might visit page 2 first, so we need to pad blank spots in the array
      for (let i = skip; i < skip + incoming.length; ++i) {
        merged[i] = incoming[i - skip];
      }
      return merged;
    },
    read(existing = [], { args }) {
      // When we fire off a query, Apollo will check the cache first. We control that with a read method
      const { skip, first } = args;
      console.log('reading...', { args });
      console.log(existing);
      // See if we have the items we want
      const items = existing
        .slice(args.skip, args.skip + args.first)
        // we filter for empty spots because its likely we have padded spots with nothing in them.
        .filter(x => x);
      // If there are items, and they aren't blank spots, return them.
      if (items.length) {
        console.log('We have items! Gonna serve them from the cache', {
          items,
        });
        return items;
      }
      // Otherwise this function returns undefined and it will hit the network for the items, and call merge() for us
    },
  };
}
