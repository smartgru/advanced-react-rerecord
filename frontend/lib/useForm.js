import { useState } from 'react';

export default function useForm(initial = {}) {
  const [inputs, updateInputs] = useState(initial);

  function handleChange(e) {
    updateInputs({
      ...inputs,
      [e.target.name]: e.target.value,
    });
  }

  function resetForm() {
    // oh - just use the default state passed in
    updateInputs(initial);
  }

  return {
    inputs,
    handleChange,
    resetForm,
  };
}
