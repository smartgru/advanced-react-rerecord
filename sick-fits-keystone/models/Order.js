const {
  Text,
  Integer,
  Relationship,
  DateTime,
  Virtual,
} = require('@keystonejs/fields');

const { byTracking, atTracking } = require('@keystonejs/list-plugins');

module.exports = {
  labelResolver: item => `Sale ${item.length}`,
  fields: {
    total: { type: Integer },
    items: {
      type: Relationship,
      ref: 'OrderItem',
      many: true,
    },
    user: {
      type: Relationship,
      ref: 'User',
    },
    charge: { type: Text },
    // TODO: Wait for virtual fields to be published
    // display_charge: {
    //   type: Virtual,
    //   resolver: item => `Charge # ${item.charge}`,
    // },
    createdAt: {
      type: DateTime,
      format: 'MM/DD/YYYY h:mm A',
      yearRangeFrom: 1901,
      yearRangeTo: 2018,
      yearPickerType: 'auto',
    },
  },
  plugins: [atTracking(), byTracking()],
};
