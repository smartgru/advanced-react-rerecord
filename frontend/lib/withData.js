import withApollo from 'next-with-apollo';
import { ApolloClient, InMemoryCache, ApolloLink } from '@apollo/client';
import { onError } from '@apollo/link-error';
import { createUploadLink } from 'apollo-upload-client';
import { getDataFromTree } from '@apollo/react-ssr';
import { endpoint, prodEndpoint } from '../config';
import paginationField from './paginationField';

function createClient({ headers, initialState }) {
  console.log('initialState');
  console.log(initialState);
  return new ApolloClient({
    link: ApolloLink.from([
      onError(({ graphQLErrors, networkError }) => {
        if (graphQLErrors)
          graphQLErrors.forEach(({ message, locations, path }) =>
            console.log(
              `[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`
            )
          );
        if (networkError) console.log(`[Network error]: ${networkError}`);
      }),
      // this uses apollo-link-http under the hood, so all the options here come from that package
      createUploadLink({
        uri: process.env.NODE_ENV === 'development' ? endpoint : prodEndpoint,
        fetchOptions: {
          credentials: 'include',
        },
        // pass the headers along from this request. This enables SSR with logged in state
        headers,
      }),
    ]),
    cache: new InMemoryCache({
      typePolicies: {
        Query: {
          fields: {
            allItems: paginationField(),
          },
        },
      },
    }).restore(initialState || {}),
  });
}

export default withApollo(createClient, { getDataFromTree });
