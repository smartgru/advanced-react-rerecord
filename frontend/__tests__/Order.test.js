import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';
import Order, { SINGLE_ORDER_QUERY } from '../components/Order';
import { fakeOrder } from '../lib/testUtils';

const mocks = [
  {
    request: { query: SINGLE_ORDER_QUERY, variables: { id: 'ord123' } },
    result: { data: { Order: fakeOrder() } },
  },
];

describe('<Order/>', () => {
  it('renders the order', async () => {
    const { container } = render(
      <MockedProvider mocks={mocks}>
        <Order id="ord123" />
      </MockedProvider>
    );
    await screen.findByTestId('order');
    expect(container).toMatchSnapshot();
  });
});
