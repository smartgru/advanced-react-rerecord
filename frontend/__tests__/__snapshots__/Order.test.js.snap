// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Order/> renders the order 1`] = `
<div>
  <div
    class="OrderStyles-sc-4oqalm-0 gjnIaU"
    data-testid="order"
  >
    <p>
      <span>
        Order ID:
      </span>
      <span>
        ord123
      </span>
    </p>
    <p>
      <span>
        Charge
      </span>
      <span>
        ch_123
      </span>
    </p>
    <p>
      <span>
        Date
      </span>
      <span>
        December 11, 2022 3:16 PM
      </span>
    </p>
    <p>
      <span>
        Order Total
      </span>
      <span>
        $400
      </span>
    </p>
    <p>
      <span>
        Item Count
      </span>
      <span>
        2
      </span>
    </p>
    <div
      class="items"
    >
      <div
        class="order-item"
      >
        <img
          src="reprehenderit.jpg"
        />
        <div
          class="item-details"
        >
          <h2>
            soluta non omnis consequatur enim quia autem
          </h2>
          <p>
            Qty: 
            1
          </p>
          <p>
            Each: 
            $42.34
          </p>
          <p>
            SubTotal: 
            $42.34
          </p>
          <p>
            et modi tenetur amet modi reprehenderit omnis
          </p>
        </div>
      </div>
      <div
        class="order-item"
      >
        <img
          src="sint.jpg"
        />
        <div
          class="item-details"
        >
          <h2>
            quia sed exercitationem omnis laborum exercitationem est
          </h2>
          <p>
            Qty: 
            1
          </p>
          <p>
            Each: 
            $42.34
          </p>
          <p>
            SubTotal: 
            $42.34
          </p>
          <p>
            sapiente laudantium molestias assumenda quasi adipisci mollitia
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
